### ISSUE TITLE
Better 'model' column handling

### ISSUE THREAD

**Issue by znicholls** [openclimatedata/openscm#114]
**Is your feature request related to a problem? Please describe.**

It's not clear how best to handle pyam's model column in OpenSCM. 

**Describe the solution you'd like**

A clean consistent way to handle the 'model' column. In order to avoid confusing users of both packages, I would suggest that we keep the model column for IAMs only, and instead use `climate_model` for the SCM which has been used to run any given scenario. When there is no 'model' associated with a given scenario (e.g. when doing an idealised 1% per year CO2 run), we would fill the model column with `None`. When plotting, if the 'model' column is `None`, we don't show that on any legends.

**Describe alternatives you've considered**

Fill the 'model' column with a different key like 'unspecified'. That may be clearer, or may just be more confusing.

**Additional context**

Requires input of pyam authors too, @danielhuppmann @gidden

**Comment by danielhuppmann** [comment]
Sounds like a good approach to me, @znicholls. I believe that `None` is as intuitive as any string like `unspecified` and provides easier handling internally.

**Comment by swillner** [comment]
Is the column `model` actually necessary in `ScmDataFrame`? Can't it just use columns it does not know as keys for an ensemble?

**Comment by znicholls** [comment]
> Is the column `model` actually necessary in `ScmDataFrame`?

Yes, to ensure that integration with pyam is as simple as possible.

> Can't it just use columns it does not know as keys for an ensemble?

I don't know what you mean here.

**Comment by swillner** [comment]
I mean, `ScmDataFrame` could just ignore columns that are note `ScmDataFrame` columns (and just "pipe them throught" so not problem with pyam). But, ignoring these columns could make the other columns non-unique (e.g. if the table contains SCM runs for several different IAMs). Thus, `ScmDataFrame` could you view this table as a collection of groups, which are each defined by their unique values of all ignore columns.

**Comment by znicholls** [comment]
I still don't fully understand. To try and give some more information, to make an `IamDataFrame` you have to include a `model` column. This is awkward for us as we'd ideally just not have it. This issue is discussing what the next best option is, the consensus is that using `None` as a fill value is a good approach.

I'm not sure where `ScmDataFrame` would ignore the `model` column but the current draft basically treats the data as timeseries with metadata, where you can put whatever you want in metadata. The catch is you have to autofill the `model` column if the user doesn't fill it.

**Comment by swillner** [comment]
So you only need to autofill the `model` column when converting to an `IamDataFrame`, right (because that needs one)? But when creating a "fresh" `ScmDataFrame` that should not be necessary, right?

**Comment by znicholls** [comment]
> So you only need to autofill the `model` column when converting to an `IamDataFrame`, right

Technically yes, but I am suggesting to autofill it always to avoid people thinking the name `model` is free when really it's not. This avoids users being confused later if they do try to work with pyam. Yes it's a price to pay, but one which is better than the alternative of behaviour suddenly changing depending on the context.

**Comment by swillner** [comment]
Ah, ok. I personally would find an unspecificed `model` column more confusing (especially since one would expect that to be the SCM). But that's probably just a matter of taste...

**Comment by znicholls** [comment]
> I personally would find an unspecificed `model` column more confusing (especially since one would expect that to be the SCM)

I think that's exactly the point. The most obvious behaviour is that `model` is the SCM. As we can't do that, we want to make the user immediately aware that the most obvious thing isn't the case. Automatically filling with `None` makes that obvious, as will a clear explanation in the docstring and the docs (which will come with the `ScmDataFrame` PR).

**Comment by znicholls** [comment]
we should keep https://github.com/IAMconsortium/pyam/issues/208 in mind too, looks like `None` may not work for us

**Comment by lewisjared** [comment]
I'm in favour of dropping the requirement on the "model" column. If it isn't present it can always be filled with a default in `ScmDataFrame.to_iam`. There is an explicit cast required to get an `IamDataFrame` which should document the caveats.

> we should keep [IAMconsortium/pyam#208](https://github.com/IAMconsortium/pyam/issues/208) in mind too, looks like `None` may not work for us

Not necessarily true. `to_iam` can convert any None's to "unspecified" or some other placeholder

**Comment by znicholls** [comment]
> I'm in favour of dropping the requirement on the "model" column.

Ok but how do we flag to users that the intuitive thing (i.e. that "model" represents "climate_model") will not work when you want to work with IAM data? Make `climate_model` a requirement?

> Not necessarily true

The reason I thought this wasn't work was a Pandas issue rather than a Pyam one. My experience with Panda's handling of None and nan in general is that it's unpleasant and you're better off using an obvious junk placeholder string than having nan.

**Comment by danielhuppmann** [comment]
Seeing that this is an ongoing unresolved discussion (and in light of recent `pyam` developments), I rescind my earlier comments and think that in an `ScmDataFrame`, the `model` should indeed be the climate model.

When passing the information back to the `IamDataFrame`, the function `to_iam()` could rename the `model` column to `climate_model` (which would be included as an extra-column in the `IamDataFrame.data` table) and add a new `model` column with `'unspecified'` or the model name which originally provided the emissions timeseries.

One could even store the name of the Integrated Assessment Model in an extra-column in the `ScmDataFrame` as `ia_model`.

**Comment by znicholls** [comment]
@danielhuppmann I wonder if that way could just cause more confusion? Given `ScmDataFrame` and `IamDataFrame` behave so similarly, having to work out whether you've got an instance of one or the other in order to infer what the `model` column means seems like more trouble than it's worth? Also how do you communicate to users who don't read the docs that `iam_model` has a special meaning and handling?

I would remain in favour of keeping `model` being IAM, `climate_model` being climate model and just filling the rows without information with `np.nan` or some other string as required (whatever we find works).

**Comment by danielhuppmann** [comment]
@znicholls, seems that there is no good option to resolve the confusion...

Maybe having a class-dependent list of required columns for the "index" may be the best way forward?
So an `IamDataFrame` requires `['model', 'scenario']` (plus region, time, unit and value), and an `ScmDataFrame` requires `['climate_model', 'scenario']`.

**Comment by znicholls** [comment]
+1 from me, seems like the least worst option as far as I can tell

**Comment by rgieseke** [comment]
@danielhuppmann @znicholls @lewisjared @swillner 

Maybe that has been discussed already and i missed it, but couldn't we call these columns `scm` and `iam`?

If it's confusing us few having `model` with potentially multiple meanings seems quite confusing.

**Comment by danielhuppmann** [comment]
> Maybe that has been discussed already and i missed it, but couldn't we call these columns `scm` and `iam`?

-1, because we have a potential user base of energy (or sector-specific) modellers, who wouldn’t consider themselves as IAMs.

**Comment by rgieseke** [comment]
> -1, because we have a potential user base of energy (or sector-specific) modellers, who wouldn’t consider themselves as IAMs.

Good point, i wasn't aware of that!
